### PR TITLE
feat: include coder_parameters from external modules

### LINKED ISSUE #8090: coder_parameters of terraform module are not properly collected
Hi,

I have created a Terraform module, `coder-template-base`, which contains all the basic resources and data elements for all my Coder workspaces in Kubernetes, since these are mostly identical for all my workspace templates. In particular, `coder-template-base` contains several `coder_parameters` in order to allow users to adjust e.g. the home volume size. 

The actual Coder workspace templates use and parameterize this module, roughly like this:

```terraform
module "coder_template_base" {
  source = "git::https://gitlab.company.com/coder-templates/coder-template-base.git?ref=1.4.0"

  image_name = var.image_name
  image_tag  = var.image_tag

  home_vol_size_default  = 5

 coder_agent_startup_script_extension = <<EOT
 echo "Hello world"
 EOT
}
```

This works fine, the Coder UI properly shows input fields for all `coder_parameters` of `coder-template-base`, as long as the templates that use this module do not define `coder_parameters` of their own. If the template defines additional `coder_parameters` like in the example below, the Coder UI only shows these additional `coder_parameters`, but not those of `coder-template-base`.

```terraform
module "coder_template_base" {
  source = "git::https://gitlab.example.com/coder-templates/coder-template-base.git?ref=1.4.0"

  image_name = var.image_name
  image_tag  = var.image_tag

  home_vol_size_default  = 5

 coder_agent_startup_script_extension = <<EOT
 curl -SL ${data.coder_parameter.tool_download_link.value} | tar -zx
 EOT
}

data "coder_parameter" "tool_download_link" {
  name         = "tool_download_link"
  display_name = "Tool Download Link"
  type         = "string"
  default      = "https://example.com/tool.tar.gz"
}
```

It looks like the algorithm, which collects all `coder_parameters` does not examine the imported modules if it already finds `coder_parameters` in the template's terraform files.

### PR DESCRIPTION
**Pull request by mtojek**
Fixes: https://github.com/coder/coder/issues/8090

This PR modifies the `ConvertResources` logic to include `coder_parameter`s from external modules. It also refactors the routine to prevent `nil` resources.

### CODE REVIEW

File: provisioner/terraform/resources.go [hunk 1]
```diff
@@ -723,29 +724,40 @@ func orderedRichParametersResources(tfResourcesRichParameters []*tfjson.StateRes
...
-				nonNil = append(nonNil, resource)
+	// Edge case: a parameter is present in an external module (Git repository, static files, etc.),
+	// which can't be easily parsed to check the parameter order.
+	// Those parameters will be prepended to the "ordered" list.
+	var external []*tfjson.StateResource
+	for _, resource := range tfResourcesRichParameters {
+		for _, o := range ordered {
+			if resource.Name == o.Name {
+				goto next
 			}
```
Comment (johnstcn) [inline provisioner/terraform/resources.go:744]: I don't have anything in particular against the `goto` here, but `continue outerLoop` is likely to raise fewer eyebrows.

**Review by johnstcn — APPROVED** [review]

File: provisioner/terraform/resources.go [hunk 2]
```diff
@@ -723,29 +724,40 @@ func orderedRichParametersResources(tfResourcesRichParameters []*tfjson.StateRes
...
 
+	if len(external) > 0 {
+		sort.Slice(external, func(i, j int) bool {
+			return external[i].Name < external[j].Name
+		})
+
+		withExternal := make([]*tfjson.StateResource, 0, len(ordered)+len(external))
+		withExternal = append(withExternal, external...)
+		withExternal = append(withExternal, ordered...)
+		ordered = withExternal
```
Comment (mafredri) [inline provisioner/terraform/resources.go:760]: This could be simplified:

```suggestion
		ordered = append(external, ordered...)
```

**Review by mafredri — APPROVED** [review]

File: provisioner/terraform/resources.go (hunk 1 above)
Comment (mtojek) [reply provisioner/terraform/resources.go]: I hear you. I replaced `goto` with a simple boolean check.

File: provisioner/terraform/resources.go (hunk 2 above)
Comment (mtojek) [reply provisioner/terraform/resources.go]: Fixed!

**Pull request merged**